bot/core/paginator.py

```diff
@@ -0,0 +1,421 @@
+import asyncio
+import logging
+import typing as t
+from textwrap import dedent
+
+import discord
+from discord import Embed
+from discord.abc import Messageable
+from discord.ext import commands
+from discord.ext.commands import Context, Paginator
+
+logger = logging.getLogger(__name__)
+
+
+class PaginatorError(Exception):
+    """
+    Exception Class for Custom error PaginatorError.
+    Raised if the bot is unable to create/process a paginator.
+    """
+    pass
+
+
+def check_permissions(channel: Messageable, bot: commands.Bot) -> bool:
+    """
+    Checks if the bot has required permissions to send or process
+    the paginator
+
+    Args:
+        channel (Messageable): The Context
+        bot (commands.Bot): The bot object, used for using bot.user if there is no guild
+
+    Returns:
+        bool: Whether the bot has all of the required permissions
+    """
+    perms: discord.Permissions = channel.permissions_for(
+        channel.guild.me if channel.guild is not None else bot.user
+    )
+
+    necessary_permissions = [
+        perms.send_messages,
+        perms.embed_links,
+        perms.add_reactions,
+        perms.read_message_history
+    ]
+
+    return all(necessary_permissions)
+
+
+class _BasePaginator:
+    """
+    An embed can be added to the embeds list
+    if the paginator class is inheriting from
+    this class
+    """
+    def __init__(self, embeds: t.List[Embed]) -> None:
+        self.embeds = embeds
+
+    def __add__(self, other: Embed) -> None:
+        self.embeds.append(other)
+
+
+class FieldPaginator(_BasePaginator):
+    """
+    A class which creates a list of embeds and configures their
+    fields according to the parameters.
+
+    Parameters
+    ----------
+        *args : t.Union[str, Embed]
+            Brief Explanation: The arguments can be a list,
+            representing a page or it can be an embed.
+
+            Example:
+            Key-Value is required. Inline is optional since its
+            set to True as a default value.
+            To insert an embed (creating a new page),
+            create a new value on the list as shown below.
+
+            FieldPaginator([{'Page': '1', 'inline': True}, discord.Embed(title="Page 2")])
```

Comment:
This should be a docstring to `__init__` function, not the class itself, also `----------` serves as sufficient separation, there is no need to also indent the actual parameters mentioned